File: javascript_code/graph/AdjSet.js


// 邻接表表达图结构
// 空间复杂度：O(V + E)
export class AdjSet {
    // lines 表示图的信息，从外面传进来
    constructor(lines) {
        let arr = lines[0]
        this.V = parseInt(arr[0])
        this.W = parseInt(arr[1])

        this.adjSet = new Array(this.V).fill(0)
        for (let i = 0; i < this.V; i++) {
            this.adjSet[i] = new Set()
        }

        for (let i = 1; i < lines.length; i++) {
            arr = lines[i]
            const a = parseInt(arr[0])
            this.validateVertex(a)
            const b = parseInt(arr[1])
            this.validateVertex(b)

            // 检测自环边
            if (a == b) {
                throw new Error("出现了自环边，错误")
            }
            // 检测平行边
            if (this.adjSet[a][b] == 1) {
                throw new Error("出现了平行边，错误")
            }

            this.adjSet[a].add(b)
            this.adjSet[b].add(a)
        }
    }

    validateVertex(v) {
        if (v < 0 || v >= this.V) {
            throw new Error("顶点 {0} 不合格".format(v))
        }
    }

    getV() {
        return this.V
    }

    getW() {
        return this.W
    }

    // 判断两个指定的顶点之间是否有边
    // 时间复杂度：O(1)
    hasEdge(v, w) {
        this.validateVertex(v)
        this.validateVertex(w)

        return this.adjSet[v].has(w)
    }

    // 获取指定顶点所有相邻的顶点
    // 时间复杂度：O(1)
    adj(v) {
        this.validateVertex(v)

        return this.adjSet[v]
    }

    // 获取指定顶点的度数
    degree(v) {
        return this.adj(v).length
    }

    toString() {
        let res = ""
        res += `顶点数 = ${this.V}，边数 = ${this.W} \n`
        for (let v = 0; v < this.V; v++) {
            res += `${v}: `
            for (const w of this.adjSet[v]) {
                res += `${w} `
            }
            res += "\n"
        }
        return res
    }
}

File: cpp_code/graph/AdjSet.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include "Graph.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <set>

class AdjSet: public Graph {
private:
    int V;
    int E;
    vector<set<int>> adjSet; // 邻接表

    void validateVertex(int v) {
        if (v < 0 || v >= V) {
            throw "顶点 " + to_string(v) + " 不合格";
        }
    }

public:
    AdjSet(string filename) {
        char c[100];
        ifstream reader(filename, ios::in | ios::binary);

        int i = 0;
        while (reader.getline(c, 100)) {
            string line = string (c);
            vector<string> arr;
            split(line, arr, " ");
            if (i == 0) {
                this->V = atoi(arr[0].c_str());
                this->E = atoi(arr[1].c_str());

                for (int i = 0; i < V; ++i) {
                    set<int> temp;
                    this->adjSet.push_back(temp);
                }
                i++;
                continue;
            }
            int a = atoi(arr[0].c_str());
            validateVertex(a);
            int b = atoi(arr[1].c_str());
            validateVertex(b);

            // 检测自环边
            if (a == b) {
                throw "出现了自环边，错误";
            }
            // 检测平行边
            if (adjSet[a].count(b)) {
                throw "出现了平行边，错误";
            }
            adjSet[a].insert(b);
            adjSet[b].insert(a);
        }
    }

    int getV() {
        return V;
    }

    int getE() {
        return E;
    }

    // 判断两个指定的顶点之间是否有边
    // 时间复杂度：O(logV)
    bool hasEdge(int v, int w) {
        validateVertex(v);
        validateVertex(w);

        return adjSet[v].count(w);
    }

    // 获取指定顶点所有相邻的顶点
    // 时间复杂度：O(V)
    vector<int> adj(int v) {
        validateVertex(v);
        vector<int> res;
        res.assign(adjSet[v].begin(), adjSet[v].end());
        return res;
    }

    // 获取指定顶点的度数
    int degree(int v) {
        return adj(v).size();
    }

    string toString() {
        stringstream fmt;
        fmt << "顶点数 = " << to_string(V) << "，边数 = " << to_string(E) << "\n";
        for (int v = 0; v < V; v++) {
            fmt << to_string(v) << ": ";
            for (auto& w : adjSet[v]) {
                fmt << to_string(w) << " ";
            }
            fmt << "\n";
        }
        string res = fmt.str();
        return res;
    }
};


/*int main() {
    Graph *g = new AdjSet("C:/workspace/douma-algo/data/graph.txt");
    cout << g->toString() << endl;
    return 0;
}*/
